calico_kubernetes/calico_kubernetes.py

```diff
@@ -141,8 +128,233 @@ def _configure_interface(self):
+            pool = None
+            for candidate_pool in pools:
+                if ip in candidate_pool:
+                    pool = candidate_pool
+                    break
+            if pool is None:
+                print("%s is not in any configured pools" % ip)
+                sys.exit(1)
+
+            # Assign the IP
+            if not self._ipam_client.assign_address(pool, ip):
+                print("IP address is already assigned in pool %s " % pool)
+                sys.exit(1)
+
+    def _unassign_addresses(self, ip_list):
+        for ip in ip_list:
+            pools = self._datastore_client.get_ip_pools(ip.version)
+            for pool in pools:
+                if ip in pool:
+                    # Ignore failure to unassign address, since we're not
+                    # enforcing assignments strictly in datastore.py.
+                    self._ipam_client.unassign_address(pool, ip)
+
+    def _create_endpoint(self, workload_id, ip, hostname, orchestrator_id):
+         # The next hop IPs for this host are stored in etcd.
+        next_hops = self._datastore_client.get_default_next_hops(hostname)
+        try:
+            next_hop = next_hops[ip.version]
+        except KeyError:
+            print("This node is not configured for IPv%d." % ip.version)
+            sys.exit(1)
+
+        network = IPNetwork(IPAddress(ip))
+        ep = Endpoint(hostname=hostname,
+                      orchestrator_id=orchestrator_id,
+                      workload_id=workload_id,
+                      endpoint_id=uuid.uuid1().hex,
+                      state="active",
+                      mac=None)
+        if network.version == 4:
+            ep.ipv4_nets.add(network)
+            ep.ipv4_gateway = next_hop
+        else:
+            ep.ipv6_nets.add(network)
+            ep.ipv6_gateway = next_hop
+
+        self._datastore_client.set_endpoint(ep)
+
+        return ep
+
+    def _create_veth(self, ep, pid, ip, interface):
+        # Create the veth, move into the container namespace, add the IP and
+        # set up the default routes.
+        netns.create_veth(ep.name, ep.temp_interface_name)
+        netns.move_veth_into_ns(pid, ep.temp_interface_name, interface)
+        netns.add_ip_to_ns_veth(pid, ip, interface)
+        if ip.version == 4:
+            netns.add_ns_default_route(pid, ep.ipv4_gateway, interface)
+        else:
+            netns.add_ns_default_route(pid, ep.ipv6_gateway, interface)
+
+        # Grab the MAC assigned to the veth in the namespace.
+        ep.mac = netns.get_ns_veth_mac(pid, interface)
+
+        # Register the endpoint with Felix.
+        self._datastore_client.set_endpoint(ep)
+
+        return ep
+
+    def _get_container_info(self, container_name):
+        try:
+            info = self._docker_client.inspect_container(container_name)
+        except APIError as e:
+            if e.response.status_code == 404:
+                print("Container %s was not found." % container_name)
```

Comment:
Remember to make it clear when we're about to exit - could lead to confusing diagnostics otherwise.